segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -636,32 +638,41 @@ private void handleWriteException(Throwable ex) {
-            } else if (ex instanceof BKException.BKLedgerClosedException) {
-                // LedgerClosed can happen because we just rolled over the ledgers or because BookKeeper closed a ledger
-                // due to some error. In either case, this is a retryable exception.
-                ex = new WriteFailureException("Active Ledger is closed.", ex);
-            } else if (ex instanceof BKException.BKWriteException) {
-                // Write-related failure or current Ledger closed. This is a retryable exception.
-                ex = new WriteFailureException("Unable to write to active Ledger.", ex);
-            } else if (ex instanceof BKException.BKClientClosedException) {
-                // The BookKeeper client was closed externally. We cannot restart it here. We should close.
-                ex = new ObjectClosedException(this, ex);
-            } else {
-                // All the other kind of exceptions go in the same bucket.
+            if (!(ex instanceof BKException)) {
                 ex = new DurableDataLogException("General exception while accessing BookKeeper.", ex);
+            } else {
+                BKException bKException = (BKException) ex;
+                switch (bKException.getCode()) {
+                    case Code.LedgerFencedException:
+                        // We were fenced out.
+                        ex = new DataLogWriterNotPrimaryException("BookKeeperLog is not primary anymore.", ex);
+                        break;
+                    case Code.NotEnoughBookiesException:
+                        // Insufficient Bookies to complete the operation. This is a retryable exception.
+                        ex = new DataLogNotAvailableException("BookKeeperLog is not available.", ex);
+                        break;
+                    case Code.LedgerClosedException:
+                        // LedgerClosed can happen because we just rolled over the ledgers or because BookKeeper closed a ledger
+                        // due to some error. In either case, this is a retryable exception.
+                        ex = new WriteFailureException("Active Ledger is closed.", ex);
+                        break;
+                    case Code.WriteException:
+                        // Write-related failure or current Ledger closed. This is a retryable exception.
+                        ex = new WriteFailureException("Unable to write to active Ledger.", ex);
+                        break;
+                    case Code.ClientClosedException:
+                        // The BookKeeper client was closed externally. We cannot restart it here. We should close.
+                        ex = new ObjectClosedException(this, ex);
+                        break;
+                    default:
+                        // All the other kind of exceptions go in the same bucket.
+                        ex = new DurableDataLogException("General exception while accessing BookKeeper.", ex);
```

Comment:
We should find a way to unify this line with 647. 
Maybe we can move this while `switch` into its own method which will return `null` if `ex` is not `BKException` of if we get in `default`. Then, back in this method, if the new method returns null, we just convert `ex` to a `DurableDataLogException`.

Reply:
@andreipaduroiu done. I have taken another approch, please see if it is looks good to your